File: .github/agents/cli-example.ts
/**
 * Example TypeScript/Node.js script for using multiple AI coding agents
 * Demonstrates CLI usage with OAuth/API key authentication
 */

import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import { GoogleGenerativeAI } from '@google/generative-ai';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config();

interface AgentConfig {
  name: string;
  apiKeyEnv: string;
  model: string;
  temperature?: number;
  maxTokens?: number;
}

class AIAgentManager {
  private configs: Record<string, AgentConfig>;
  private openaiClient: OpenAI | null = null;
  private anthropicClient: Anthropic | null = null;
  private geminiClient: GoogleGenerativeAI | null = null;

  constructor() {
    this.configs = {
      openai: {
        name: 'OpenAI Codex',
        apiKeyEnv: 'OPENAI_API_KEY',
        model: 'gpt-4-turbo-preview',
        temperature: 0.7,
        maxTokens: 2048,
      },
      anthropic: {
        name: 'Anthropic Claude',
        apiKeyEnv: 'ANTHROPIC_API_KEY',
        model: 'claude-3-sonnet-20240229',
        temperature: 0.7,
        maxTokens: 4096,
      },
      google: {
        name: 'Google Gemini',
        apiKeyEnv: 'GOOGLE_API_KEY',
        model: 'gemini-pro',
        temperature: 0.7,
        maxTokens: 2048,
      },
    };

    this.setupClients();
  }

  private setupClients(): void {
    // OpenAI
    const openaiKey = process.env[this.configs.openai.apiKeyEnv];
    if (openaiKey) {
      this.openaiClient = new OpenAI({ apiKey: openaiKey });
      console.log(`✓ ${this.configs.openai.name} initialized`);
    } else {
      console.log(
        `⚠ ${this.configs.openai.name} not configured (missing ${this.configs.openai.apiKeyEnv})`
      );
    }

    // Anthropic
    const anthropicKey = process.env[this.configs.anthropic.apiKeyEnv];
    if (anthropicKey) {
      this.anthropicClient = new Anthropic({ apiKey: anthropicKey });
      console.log(`✓ ${this.configs.anthropic.name} initialized`);
    } else {
      console.log(
        `⚠ ${this.configs.anthropic.name} not configured (missing ${this.configs.anthropic.apiKeyEnv})`
      );
    }

    // Google Gemini
    const geminiKey = process.env[this.configs.google.apiKeyEnv];
    if (geminiKey) {
      this.geminiClient = new GoogleGenerativeAI(geminiKey);
      console.log(`✓ ${this.configs.google.name} initialized`);
    } else {
      console.log(
        `⚠ ${this.configs.google.name} not configured (missing ${this.configs.google.apiKeyEnv})`
      );
    }
  }

  async generateCodeOpenAI(prompt: string): Promise<string | null> {
    if (!this.openaiClient) return null;

    try {
      const response = await this.openaiClient.chat.completions.create({
        model: this.configs.openai.model,
        messages: [
          { role: 'system', content: 'You are an expert coding assistant.' },
          { role: 'user', content: prompt },
        ],
        temperature: this.configs.openai.temperature,
        max_tokens: this.configs.openai.maxTokens,
      });

      return response.choices[0]?.message?.content || null;
    } catch (error) {
      console.error(`Error with OpenAI:`, error);
      return null;
    }
  }

  async generateCodeAnthropic(prompt: string): Promise<string | null> {
    if (!this.anthropicClient) return null;

    try {
      const message = await this.anthropicClient.messages.create({
        model: this.configs.anthropic.model,
        max_tokens: this.configs.anthropic.maxTokens || 4096,
        messages: [{ role: 'user', content: prompt }],
      });

      return message.content[0].type === 'text' ? message.content[0].text : null;
    } catch (error) {
      console.error(`Error with Anthropic:`, error);
      return null;
    }
  }

  async generateCodeGemini(prompt: string): Promise<string | null> {
    if (!this.geminiClient) return null;

    try {
      const model = this.geminiClient.getGenerativeModel({
        model: this.configs.google.model,
      });

      const result = await model.generateContent(prompt);
      const response = await result.response;
      return response.text();
    } catch (error) {
      console.error(`Error with Gemini:`, error);
      return null;
    }
  }

  async generateCode(prompt: string, agent: string): Promise<string | null> {
    const generators: Record<string, (p: string) => Promise<string | null>> = {
      openai: this.generateCodeOpenAI.bind(this),
      anthropic: this.generateCodeAnthropic.bind(this),
      google: this.generateCodeGemini.bind(this),
    };

    const generator = generators[agent];
    if (!generator) {
      console.error(`Unknown agent: ${agent}`);
      return null;
    }

    return generator(prompt);
  }

  async compareAgents(prompt: string): Promise<Record<string, string | null>> {
    const results: Record<string, string | null> = {};

    console.log('\n' + '='.repeat(80));
    console.log('Comparing AI Agents');
    console.log('='.repeat(80));

    for (const agentName of ['openai', 'anthropic', 'google']) {
      console.log(`\nGenerating with ${this.configs[agentName].name}...`);
      const result = await this.generateCode(prompt, agentName);
      results[agentName] = result;

      if (result) {
        console.log(`✓ Response received (${result.length} characters)`);
      } else {
        console.log(`✗ No response`);
      }
    }

    return results;
  }
}

// CLI Implementation
async function main() {
  const args = process.argv.slice(2);
  
  // Simple argument parsing
  const getArg = (flag: string): string | undefined => {
    const index = args.indexOf(flag);
    return index !== -1 && args[index + 1] ? args[index + 1] : undefined;
  };
  
  const hasFlag = (flag: string): boolean => args.includes(flag);

  if (hasFlag('--help') || hasFlag('-h')) {
    console.log(`
AI Coding Agent CLI (TypeScript)

Usage:
  ts-node cli-example.ts [options]

Options:
  --agent <name>      AI agent to use (openai, anthropic, google) [default: openai]
  --prompt <text>     Prompt for code generation
  --compare           Compare responses from all agents
  --help, -h          Show this help message

Examples:
  # Generate code with OpenAI
  ts-node cli-example.ts --agent openai --prompt "Write a TypeScript function to sort an array"
  
  # Compare responses from all agents
  ts-node cli-example.ts --compare --prompt "Write a REST API handler"

Environment Variables:
  OPENAI_API_KEY      - OpenAI API key
  ANTHROPIC_API_KEY   - Anthropic API key
  GOOGLE_API_KEY      - Google API key
    `);
    process.exit(0);
  }

  const agent = getArg('--agent') || 'openai';
  const prompt = getArg('--prompt');
  const compare = hasFlag('--compare');

  if (!prompt) {
    console.error('Error: --prompt is required');
    process.exit(1);
  }

  const manager = new AIAgentManager();

  if (compare) {
    const results = await manager.compareAgents(prompt);

    console.log('\n' + '='.repeat(80));
    console.log('RESULTS');
    console.log('='.repeat(80));

    for (const [agentName, result] of Object.entries(results)) {
      console.log(`\n## ${manager['configs'][agentName].name}\n`);
      if (result) {
        console.log(result);
      } else {
        console.log('[No response]');
      }
      console.log('\n' + '-'.repeat(80));
    }
  } else {
    const result = await manager.generateCode(prompt, agent);

    if (result) {
      console.log(result);
    } else {
      console.error(`Failed to generate code with ${agent}`);
      process.exit(1);
    }
  }
}

// Run if called directly
if (require.main === module) {
  main().catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

export { AIAgentManager };

File: .github/agents/cli-example.py
#!/usr/bin/env python3
"""
Example script for generating code with multiple AI agents
Demonstrates CLI usage with API key authentication
"""

import os
import sys
from typing import Optional, Dict, Any
from dataclasses import dataclass

# Import AI SDKs
try:
    from openai import OpenAI
    from anthropic import Anthropic
    import google.generativeai as genai
except ImportError as e:
    print(f"Error: Missing required SDK. Run: pip install -r .github/agents/requirements.txt")
    print(f"Details: {e}")
    sys.exit(1)


@dataclass
class AgentConfig:
    """Configuration for an AI agent"""
    name: str
    api_key_env: str
    model: str
    temperature: float = 0.7
    max_tokens: int = 2048


class AIAgentManager:
    """Manager for multiple AI coding agents"""
    
    def __init__(self):
        self.configs = {
            'openai': AgentConfig(
                name='OpenAI Codex',
                api_key_env='OPENAI_API_KEY',
                model='gpt-4-turbo-preview',
                temperature=0.7,
                max_tokens=2048
            ),
            'anthropic': AgentConfig(
                name='Anthropic Claude',
                api_key_env='ANTHROPIC_API_KEY',
                model='claude-3-sonnet-20240229',
                temperature=0.7,
                max_tokens=4096
            ),
            'google': AgentConfig(
                name='Google Gemini',
                api_key_env='GOOGLE_API_KEY',
                model='gemini-pro',
                temperature=0.7,
                max_tokens=2048
            )
        }
        
        # Initialize clients
        self.openai_client = None
        self.anthropic_client = None
        self.gemini_configured = False
        
        self._setup_clients()
    
    def _setup_clients(self):
        """Initialize API clients with environment variables"""
        
        # OpenAI
        if os.getenv(self.configs['openai'].api_key_env):
            self.openai_client = OpenAI(
                api_key=os.getenv(self.configs['openai'].api_key_env)
            )
            print(f"✓ {self.configs['openai'].name} initialized")
        else:
            print(f"⚠ {self.configs['openai'].name} not configured (missing {self.configs['openai'].api_key_env})")
        
        # Anthropic
        if os.getenv(self.configs['anthropic'].api_key_env):
            self.anthropic_client = Anthropic(
                api_key=os.getenv(self.configs['anthropic'].api_key_env)
            )
            print(f"✓ {self.configs['anthropic'].name} initialized")
        else:
            print(f"⚠ {self.configs['anthropic'].name} not configured (missing {self.configs['anthropic'].api_key_env})")
        
        # Google Gemini
        if os.getenv(self.configs['google'].api_key_env):
            genai.configure(api_key=os.getenv(self.configs['google'].api_key_env))
            self.gemini_configured = True
            print(f"✓ {self.configs['google'].name} initialized")
        else:
            print(f"⚠ {self.configs['google'].name} not configured (missing {self.configs['google'].api_key_env})")
    
    def generate_code_openai(self, prompt: str) -> Optional[str]:
        """Generate code using OpenAI"""
        if not self.openai_client:
            return None
        
        try:
            response = self.openai_client.chat.completions.create(
                model=self.configs['openai'].model,
                messages=[
                    {"role": "system", "content": "You are an expert coding assistant."},
                    {"role": "user", "content": prompt}
                ],
                temperature=self.configs['openai'].temperature,
                max_tokens=self.configs['openai'].max_tokens
            )
            return response.choices[0].message.content
        except Exception as e:
            print(f"Error with OpenAI: {e}")
            return None
    
    def generate_code_anthropic(self, prompt: str) -> Optional[str]:
        """Generate code using Anthropic Claude"""
        if not self.anthropic_client:
            return None
        
        try:
            message = self.anthropic_client.messages.create(
                model=self.configs['anthropic'].model,
                max_tokens=self.configs['anthropic'].max_tokens,
                messages=[
                    {"role": "user", "content": prompt}
                ]
            )
            return message.content[0].text
        except Exception as e:
            print(f"Error with Anthropic: {e}")
            return None
    
    def generate_code_gemini(self, prompt: str) -> Optional[str]:
        """Generate code using Google Gemini"""
        if not self.gemini_configured:
            return None
        
        try:
            model = genai.GenerativeModel(self.configs['google'].model)
            response = model.generate_content(prompt)
            return response.text
        except Exception as e:
            print(f"Error with Gemini: {e}")
            return None
    
    def generate_code(self, prompt: str, agent: str = 'openai') -> Optional[str]:
        """Generate code using specified agent"""
        generators = {
            'openai': self.generate_code_openai,
            'anthropic': self.generate_code_anthropic,
            'google': self.generate_code_gemini
        }
        
        generator = generators.get(agent)
        if not generator:
            print(f"Unknown agent: {agent}")
            return None
        
        return generator(prompt)
    
    def compare_agents(self, prompt: str) -> Dict[str, Optional[str]]:
        """Compare responses from all available agents"""
        results = {}
        
        print("\n" + "="*80)
        print("Comparing AI Agents")
        print("="*80)
        
        for agent_name in ['openai', 'anthropic', 'google']:
            print(f"\nGenerating with {self.configs[agent_name].name}...")
            result = self.generate_code(prompt, agent_name)
            results[agent_name] = result
            
            if result:
                print(f"✓ Response received ({len(result)} characters)")
            else:
                print(f"✗ No response")
        
        return results


def main():
    """Main CLI function"""
    import argparse
    
    parser = argparse.ArgumentParser(
        description='AI Coding Agent CLI',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Generate code with OpenAI
  %(prog)s --agent openai --prompt "Write a Python function to sort a list"
  
  # Compare responses from all agents
  %(prog)s --compare --prompt "Write a REST API handler"
  
  # Use from stdin
  echo "Write a binary search function" | %(prog)s --agent anthropic --stdin
  
Environment Variables:
  OPENAI_API_KEY      - OpenAI API key
  ANTHROPIC_API_KEY   - Anthropic API key
  GOOGLE_API_KEY      - Google API key
        """
    )
    
    parser.add_argument('--agent', choices=['openai', 'anthropic', 'google'],
                       default='openai', help='AI agent to use')
    parser.add_argument('--prompt', type=str, help='Prompt for code generation')
    parser.add_argument('--compare', action='store_true',
                       help='Compare responses from all agents')
    parser.add_argument('--stdin', action='store_true',
                       help='Read prompt from stdin')
    parser.add_argument('--output', type=str, help='Output file (default: stdout)')
    
    args = parser.parse_args()
    
    # Get prompt
    if args.stdin:
        prompt = sys.stdin.read()
    elif args.prompt:
        prompt = args.prompt
    else:
        parser.print_help()
        print("\nError: Either --prompt or --stdin is required")
        sys.exit(1)
    
    # Initialize manager
    manager = AIAgentManager()
    
    # Generate code
    if args.compare:
        results = manager.compare_agents(prompt)
        
        print("\n" + "="*80)
        print("RESULTS")
        print("="*80)
        
        for agent, result in results.items():
            print(f"\n## {manager.configs[agent].name}\n")
            if result:
                print(result)
            else:
                print("[No response]")
            print("\n" + "-"*80)
    else:
        result = manager.generate_code(prompt, args.agent)
        
        if result:
            if args.output:
                with open(args.output, 'w') as f:
                    f.write(result)
                print(f"Output written to {args.output}")
            else:
                print(result)
        else:
            print(f"Failed to generate code with {args.agent}")
            sys.exit(1)


if __name__ == '__main__':
    main()
